File: apps/android/feature/canvas/data/src/main/java/com/scribblefit/feature/canvas/data/repository/CanvasRepositoryImpl.kt
package com.scribblefit.feature.canvas.data.repository

import com.scribblefit.feature.ai.domain.engine.SyncRepository
import com.scribblefit.feature.ai.domain.model.ParsedWorkout
import com.scribblefit.feature.ai.domain.model.SyncStatus
import com.scribblefit.feature.canvas.domain.model.FeedItem
import com.scribblefit.feature.canvas.domain.model.ScribbleStatus
import com.scribblefit.feature.canvas.domain.repository.CanvasRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.serialization.json.Json
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CanvasRepositoryImpl @Inject constructor(
    private val syncRepository: SyncRepository,
    private val json: Json
) : CanvasRepository {

    override fun getFeed(): Flow<List<FeedItem>> =
        syncRepository.getAllSyncItems().map { items ->
            items.mapNotNull { item ->
                when (item.type) {
                    "SCRIBBLE" -> {
                        if (item.status == SyncStatus.COMPLETED && item.jsonData != null) {
                            val workout = runCatching {
                                json.decodeFromString<ParsedWorkout>(item.jsonData!!)
                            }.getOrNull()
                            if (workout != null) {
                                FeedItem.Confirmation(
                                    id = item.id,
                                    timestamp = item.createdAt,
                                    workout = workout,
                                    scribbleId = item.id
                                )
                            } else {
                                FeedItem.Scribble(
                                    id = item.id,
                                    timestamp = item.createdAt,
                                    rawText = item.rawText ?: "",
                                    status = item.status.toScribbleStatus()
                                )
                            }
                        } else {
                            FeedItem.Scribble(
                                id = item.id,
                                timestamp = item.createdAt,
                                rawText = item.rawText ?: "",
                                status = item.status.toScribbleStatus()
                            )
                        }
                    }
                    else -> null
                }
            }.sortedBy { it.timestamp() }
        }

    override suspend fun addScribble(rawText: String) {
        syncRepository.enqueueScribble(UUID.randomUUID().toString(), rawText)
    }

    override suspend fun retryScribble(id: String) {
        syncRepository.updateSyncStatus(id, SyncStatus.PENDING)
        syncRepository.syncWorkouts()
    }

    override suspend fun addConfirmation(item: FeedItem.Confirmation) {
        val encoded = json.encodeToString(item.workout)
        syncRepository.saveFeedItem(item.id, "CONFIRMATION", encoded, SyncStatus.COMPLETED)
    }

    override suspend fun addInsight(item: FeedItem.Insight) {
        syncRepository.saveFeedItem(item.id, "INSIGHT", "", SyncStatus.COMPLETED)
    }

    override suspend fun removeFeedItem(id: String) {
        syncRepository.deleteSyncItem(id)
    }

    override suspend fun clearFeed() {
        // Clears handled at database level
    }

    private fun SyncStatus.toScribbleStatus(): ScribbleStatus = when (this) {
        SyncStatus.PENDING -> ScribbleStatus.PENDING
        SyncStatus.PROCESSING -> ScribbleStatus.PROCESSING
        SyncStatus.FAILED -> ScribbleStatus.FAILED
        SyncStatus.COMPLETED -> ScribbleStatus.COMPLETED
    }

    private fun FeedItem.timestamp(): Long = when (this) {
        is FeedItem.Prompt -> timestamp
        is FeedItem.Scribble -> timestamp
        is FeedItem.Confirmation -> timestamp
        is FeedItem.Insight -> timestamp
    }
}

File: apps/ios/ScribbleFit/Sources/Features/Canvas/Data/Repository/CanvasRepositoryImpl.swift
import Combine
import Foundation

public final class CanvasRepositoryImpl: CanvasRepository {
    private let syncRepository: any SyncRepository
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    public init(syncRepository: any SyncRepository) {
        self.syncRepository = syncRepository
    }

    public func getFeed() -> AnyPublisher<[FeedItem], Never> {
        syncRepository.observeAllSyncItems()
            .map { items in
                items.compactMap { item -> FeedItem? in
                    switch item.itemType {
                    case "SCRIBBLE":
                        if item.status == .completed,
                           let data = item.jsonData?.data(using: .utf8),
                           let workout = try? JSONDecoder().decode(ParsedWorkout.self, from: data) {
                            return .confirmation(ConfirmationItem(
                                id: item.id, timestamp: item.createdAt,
                                workout: workout, scribbleId: item.id
                            ))
                        }
                        return .scribble(ScribbleItem(
                            id: item.id, timestamp: item.createdAt,
                            rawText: item.rawText ?? "",
                            status: item.status.toScribbleStatus()
                        ))
                    default:
                        return nil
                    }
                }
                .sorted { $0.timestamp < $1.timestamp }
            }
            .eraseToAnyPublisher()
    }

    public func addScribble(rawText: String) async throws {
        try await syncRepository.enqueueScribble(id: UUID().uuidString, rawText: rawText)
    }

    public func retryScribble(id: String) async throws {
        try await syncRepository.updateSyncStatus(id: id, status: .pending)
        await syncRepository.syncWorkouts()
    }

    public func addConfirmation(item: ConfirmationItem) async throws {
        let data = try encoder.encode(item)
        let json = String(data: data, encoding: .utf8) ?? ""
        try await syncRepository.saveFeedItem(id: item.id, type: "CONFIRMATION", jsonData: json, status: .completed)
    }

    public func addInsight(item: InsightItem) async throws {
        let data = try encoder.encode(item)
        let json = String(data: data, encoding: .utf8) ?? ""
        try await syncRepository.saveFeedItem(id: item.id, type: "INSIGHT", jsonData: json, status: .completed)
    }

    public func removeFeedItem(id: String) async throws {
        try await syncRepository.deleteSyncItem(id: id)
    }

    public func clearFeed() async throws {}
}

private extension SyncStatus {
    func toScribbleStatus() -> ScribbleStatus {
        switch self {
        case .pending: return .pending
        case .processing: return .processing
        case .failed: return .failed
        case .completed: return .completed
        }
    }
}
